presto-main/src/main/java/io/prestosql/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -575,6 +573,50 @@ protected Type visitSimpleCaseExpression(SimpleCaseExpression node, StackableAst
             return type;
         }
 
+        private Type coerceCaseOperandToToSingleType(SimpleCaseExpression node, StackableAstVisitorContext<Context> context)
+        {
+            Type operandType = process(node.getOperand(), context);
```

Comment:
This implementation can be simplified a bit. Here's what I came up with after playing with the code a bit:

```java
Type operandType = process(node.getOperand(), context);
List<Type> whenOperandTypes = new ArrayList<>();

Type globalCommonType = null;
for (WhenClause whenClause : node.getWhenClauses()) {
    Type type = process(whenClause.getOperand(), context);
    whenOperandTypes.add(type);

    Optional<Type> commonType = typeCoercion.getCommonSuperType(operandType, type);

    if (!commonType.isPresent()) {
        throw semanticException(TYPE_MISMATCH, whenClause.getOperand(), "CASE operand type does not match WHEN clause operand type: %s vs %s", operandType, type);
    }

    if (globalCommonType == null) {
        globalCommonType = commonType.get();
    }

    globalCommonType = typeCoercion.getCommonSuperType(globalCommonType, commonType.get()).get();
}

coerceType(node.getOperand(), operandType, globalCommonType, "CASE operand");
for (int i = 0; i < node.getWhenClauses().size(); i++) {
    coerceType(node.getWhenClauses().get(i).getOperand(), whenOperandTypes.get(i), globalCommonType, "WHEN clause operand");
}

return operandType;
```